### PR TITLE
Fix iOS caching + add iOS feature: preCaching

### PR DESCRIPTION
**Pull request by themadmrj**
On iOS, caching currently does not work as expected, because the cacheKey is ignored. This PR builds a new caching system on iOS which fixes this and also allows pre-caching, which I had previously only added on Android. With this PR, betterplayer will be the only library that fully supports caching and pre-caching videos on both platforms.

### CODE REVIEW

**Comment by jhomlala** [comment]
@themadmrj Does this cache work with HLS?

**Comment by themadmrj** [comment]
The current system is to load everything into memory and save the file to disk once downloaded, so if you plan to read a stream for 2 hours, probably not. But anyway, who uses cache with hls?

**Comment by jhomlala** [comment]
@themadmrj iOS structure has been changed in last version. Could you recheck your code and fit it to the current state?

**Comment by jhomlala** [comment]
@themadmrj Could you please resolve these conflicts when you will have some spare time?

**Comment by themadmrj** [comment]
Done

**Comment by erykkruk** [comment]
Hey, Do you have plans to merge it?

**Comment by cesariomartins** [comment]
Hi,
Pls, let me know when this goes to production. I'd love to test this feature my app. many tks!!

**Comment by shilangyu** [comment]
Would you mind rebasing? The PR diff shows much more than what was changed which makes it hard to follow. Thanks

**Comment by shilangyu** [comment]
@themadmrj How hard would it be to convert this code to work as persistent downloads? Am working on downloads for android in #665, and it would be nice if it could work on iOS as well. The main differences between downloads and caching are:

- Downloads are persistent, cannot be cleared by a simple "clear cache", treated as app data
- Downloads can be queried, ie a list of downloaded files can be retrieved
- Download progress can be listened to

Thanks
